### ISSUE TITLE
Pub/Sub 0.32.0 | Exception "google.api_core.exceptions.Unknown: None Stream removed"

### ISSUE THREAD

**Issue by PicardParis** [GoogleCloudPlatform/google-cloud-python#4989]
- Context
  ```shell
  google-cloud-pubsub (0.32.0)
  Python 2.7.13
  Win10
  ```

- Subscriber console | Fails w/ exception after a random time (from 15 min to 3 h)
  ```shell
  > python pubsub_subscribe.py
  Subscribed to messages on projects/PROJECT_ID/subscriptions/TOPIC
  Waiting for future...
  Messages received: 0 + 300
  Messages received: 300 + 3000
  Traceback (most recent call last):
    File ".\pubsub_subscribe.py", line 58, in <module>
      subscription.future.result()
    File "C:\Python27\lib\site-packages\google\cloud\pubsub_v1\futures.py", line 114, in result
      raise err
  google.api_core.exceptions.Unknown: None Stream removed
  ```

- Publisher console | Completes as expected
  ```shell
  > python pubsub_publish.py
  Messages sent: 0 + 300
  Messages sent: 300 + 3000
  Messages sent: 3300 + 30000
  Messages sent: 33300 + 30
  Messages sent: 33330 + 5
  Messages sent: 33335 + 3
  Messages sent: 33338 + 2
  Messages sent: 33340 + 30000
  Messages sent: 63340 + 3000
  Messages sent: 66340 + 300
  Total messages sent: 66640
  ```

- Initialization
  - GCP project `GOOGLE_CLOUD_PROJECT`
  - Pub/Sub API enabled
  - Existing topic `PUBSUB_TOPIC`
  - Environment variables `GOOGLE_CLOUD_PROJECT` & `PUBSUB_TOPIC`

- Subscriber sample code `<pubsub_subscribe.py>`
  ```python
  import os
  from socket import gethostname
  from time import sleep
  from google.cloud import pubsub
  from google.api_core.exceptions import AlreadyExists

  PROJECT_ID = os.getenv('GOOGLE_CLOUD_PROJECT')
  TOPIC = os.getenv('PUBSUB_TOPIC')

  def on_pubsub_message(msg):
      # These cases are not handled:
      # - A message may be received more than once
      # - A message may be received out of order
      cmd = msg.data
      msg.ack()
      # print('.. msg: %s' % cmd)

      global msg_received_total
      global msg_received_loop
      global subscription

      if cmd == 'stop':
          print("Received 'stop'")
          msg_received_total += msg_received_loop
          subscription.future.set_result(True)
          return

      msg_received_loop += 1
      if cmd == 'loop_end':
          print('Messages received: %d + %d' % (
              msg_received_total, msg_received_loop))
          msg_received_total += msg_received_loop
          msg_received_loop = 0

  msg_received_total = 0
  msg_received_loop = 0

  subscriber = pubsub.SubscriberClient()
  topic_path = subscriber.topic_path(PROJECT_ID, TOPIC)
  subsc_path = subscriber.subscription_path(PROJECT_ID, gethostname())

  try:
      subscriber.create_subscription(subsc_path, topic_path)
  except AlreadyExists:
      pass

  subscription = subscriber.subscribe(subsc_path, on_pubsub_message)
  print('Subscribed to messages on %s' % subsc_path)

  print('Waiting for future...')
  subscription.future.result()

  print('Total messages received: %d' % msg_received_total)
  ```

- Publisher sample code `<pubsub_publish.py>`
  ```python
  import os
  from time import sleep
  from google.cloud import pubsub

  PROJECT_ID = os.getenv('GOOGLE_CLOUD_PROJECT')
  TOPIC = os.getenv('PUBSUB_TOPIC')

  PARAMS = [
      # (number_of_messages, seconds_after_each_message)
      (300, 1),  # 5 min
      (3000, 0.1),  # 5 min
      (30000, 0.01),  # 5 min
      (30, 10),  # 5 min
      (5, 60),  # 5 min
      (3, 600),  # 30 min
      (2, 3600),  # 120 min
      (30000, 0.01),  # 5 min
      (3000, 0.1),  # 5 min
      (300, 1),  # 5 min
  ]

  publisher = pubsub.PublisherClient()
  topic_path = publisher.topic_path(PROJECT_ID, TOPIC)

  msg_sent = 0
  for messages, delay in PARAMS:
      for i in xrange(1, messages):
          publisher.publish(topic_path, 'msg#{}'.format(i).encode('utf-8'))
          sleep(delay)
      # Keep last message to send a 'loop_end' command
      sleep(1)
      publisher.publish(topic_path, b'loop_end')
      sleep(1)
      print('Messages sent: %d + %d' % (msg_sent, messages))
      msg_sent += messages

  # publisher.publish(topic_path, b'stop')
  print('Total messages sent: %d' % msg_sent)
  ```

**Comment by jonparrott** [comment]
@PicardParis is this on Raspberry Pi?

**Comment by PicardParis** [comment]
With version 0.32.0, for the time being, I have only produced this exact issue on Win10 but previously had the same issues on both envs. I'm currently updating a Raspberry Pi and will run the same subscriber code...

**Comment by jadekler** [comment]
Hi @PicardParis. Could you provide your `pip list`?

**Comment by PicardParis** [comment]
I haven't been getting consistent results between Windows & Raspbian (didn't get the exception so far) and didn't keep the original environments. Here is the log extract I kept on Windows:
```
> pip install --upgrade google-cloud-pubsub
...
Successfully installed google-cloud-pubsub-0.32.0 psutil-5.4.3
```

Right now, on the Pi, if I query outdated modules, I get this
```shell
> pip list -o --format=columns
Package         Version Latest Type
--------------- ------- ------ -----
google-api-core 0.1.4   1.0.0  wheel
```
Might it be the issue?

I'm changing the sample code to detect potentially missed msgs and will run new tests.

**Comment by jadekler** [comment]
Could you post the version of grpc you're using here?

**Comment by jadekler** [comment]
Also just to confirm, you're getting errors on the raspberry pi but not on windows?

**Comment by PicardParis** [comment]
grpc info

```shell
grpc-google-iam-v1 (0.11.4)
grpcio (1.10.0)
```

Tests so far

Date | Publisher | Suscriber | Suscriber Power | Suscriber Network | Result | Repro#
-- | -- | -- | -- | -- | -- | --
2018-03-05 | Win10 | =Publisher | AC | Ethernet | google.api_core.exceptions | x2
2018-03-07 | Win10 | Raspbian | AC | Wi-Fi | no exception but doubt on # msgs received | x1
2018-03-07 | Win10 | Raspbian | AC | Ethernet | no exception but doubt on # msgs received | x1
2018-03-07 | Win10 | =Publisher | Battery | Wi-Fi | OK | x1

Tests are not consistent. No conclusion fttb.

**Comment by jadekler** [comment]
Thanks. I'll see if I can snag a windows laptop and repro. Let me know if you find out anything else - and again, thanks for helping debug.

**Comment by PicardParis** [comment]
Sure. Inconsistent results are pretty annoying.

**Comment by jadekler** [comment]
Hi @PicardParis. Thanks for this repro, much appreciated. I wasn't able to get a windows laptop yesterday, and am out of the office today. However! We have two raspberry pi's rush ordered _and_ should have a windows laptop on monday, so hopefully I can start digging into this soon. Sorry for the delay, hope we can figure out what's going on soon!

**Comment by jadekler** [comment]
In the meantime @PicardParis , if you could try running your app with `GRPC_VERBOSITY=debug` and `GRPC_TRACE=all`, I'd really love to see a log of what's happening (gist.github.com or whatever works).

**Comment by PicardParis** [comment]
The error could be reproduced with traces. See [gist](https://gist.github.com/PicardParis/3112de307cd2c369fdfcaa3155b7c59e).

**Comment by jonparrott** [comment]
Okay, after some discussion the proper thing for us to do here is retry on `UNKNOWN` status.

**Comment by patrickbeekman** [comment]
I was seeing the same error messages, but it was caused by a different bug with the latest `google-api-core==1.17.0`. This issue was resolved for me by downgrading to `1.16.0`. https://github.com/googleapis/python-api-core/issues/25 I hope this is helpful.
